### ISSUE TITLE
Proposal: Zipper movement API should handle nil wherever zippers are accepted

### ISSUE THREAD

**Issue by zachallaun** [doorgan/sourceror#166]
Apologies if this has been discussed/discarded before; if it has, please feel free to close!

I believe that nil-handling is the least ergonomic part of the Zipper API. For example:

```elixir
iex(1)> ":foo" |> Sourceror.parse_string!() |> Sourceror.Zipper.zip() |> Sourceror.Zipper.up()
nil
iex(2)> ":foo" |> Sourceror.parse_string!() |> Sourceror.Zipper.zip() |> Sourceror.Zipper.up() |> Sourceror.Zipper.up()
** (FunctionClauseError) no function clause matching in Sourceror.Zipper.up/1    
    
    The following arguments were given to Sourceror.Zipper.up/1:
    
        # 1
        nil
    
    Attempted function clauses (showing 2 out of 2):
    
        def up(%Sourceror.Zipper{path: nil})
        def up(%Sourceror.Zipper{node: tree, path: path, supertree: supertree})
    
    (sourceror 1.6.0) lib/sourceror/zipper.ex:208: Sourceror.Zipper.up/1
    iex:2: (file)
```

The API lends itself *incredibly* well to chaining using the pipe operator, but in almost all cases (or maybe all cases), when a zipper operation returns nil, the next operation is likely to raise a function-clause error.

I'd love to discuss adjusting the Zipper *movement* API (not necessarily the update/modification API) to "pass-through" nil when it's encountered, instead of raising.

**Comment by zachdaniel** [comment]
I'm here for it. Currently, it just blows up, so I don't think there is really any reasonable harm in adjusting it to pass through `nil`.

**Comment by doorgan** [comment]
This would be a welcome change. The `nil` handling is the result of basing the implementation on clujure/zip, which returns `nil` in those cases. I can see how that is inconvenient, so I'm very in favor of changing that behavior

**Comment by zachallaun** [comment]
Great!

I'll try to submit a PR for this change soon, unless someone else gets to it first. 🙂

**Comment by zachallaun** [comment]
Here's the list of functions that I think should accept `nil` and handle it appropriately. Basically, I made note of every function that would reasonably be used to "find forwards" in a zipper, the idea being that "finding the right thing" is often a multi-step, chained operation, and you don't care which one returned the `nil`.

In order of appearance in the docs:

```
down/1
find/3
find_all/3
find_value/3
left/1
leftmost/1
move_to_curssor/2
next/1
right/1
rightmost/1
search_pattern/2
skip/2
traverse/2
traverse/3
traverse_while/2
traverse_while/3
```

Notable exclusions:

* Functions that move "back", such as `up/1`, `top/1`, `prev/1`, etc. This is because you always want this to be true: `zipper == zipper |> Zipper.next() |> Zipper.prev()`, and it wouldn't be if `next` returned `nil` and `prev` passed the `nil` through.
* Functions that create/update/modify zippers. You should check that your "find" was successful prior to this.

Anything I missed?

**Comment by zachallaun** [comment]
As I'm getting into the implementation, I'm thinking of striking traversal off that list. It doesn't seem useful to return `{nil, acc}` for the `traverse/3` and `traverse_while/3` cases, and it seems maybe weird for the arity-2 versions to accept `nil` but not the arity-3 versions.

**Comment by doorgan** [comment]
@zachallaun that list looks good, and I agree with your reasoninig in the exclusion list

> As I'm getting into the implementation, I'm thinking of striking traversal off that list. It doesn't seem useful to return {nil, acc} for the traverse/3 and traverse_while/3 cases, and it seems maybe weird for the arity-2 versions to accept nil but not the arity-3 versions.

Yeah that makes sense, you are kinda forced to handle the possible nil result, so there's not a good way to make it "more pipeable"

**Comment by doorgan** [comment]
Merged with #167
